proposal to solve issue #74 (#78)
kg-construct/rml-core: spec/docs/joinconditions.md

* a [child map]() (`rml:childMap`),
  whose value is an [Expression Map]() (`rml:ExpressionMap`), which
  MUST include references that exists in the [Logical Source]()

>   of the [Triples Map]() that contains the [Referencing Object Map]()

Reviewer: This correction makes sense, but the whole sentence doesn't IMO.

I don't think we can say that references _exists_ in a logical source. And I also don't think it's a MUST. 

I think the sentence should be something along the lines of:

> * a child map property (`rr:childMap`) whose value is a child map (`rml:ChildMap`).

And then there should be explanation about what a `rml:ChildMap` is.
Whether or not the child map expression resolves or not is not really a concern for the spec IMO.

    rml:logicalSource <LS2> ;
    rml:subjectMap <#SM2> .
```

> ## Join types

Reviewer: I would prefer not to call this Join types. I also don't see the need for a separate section, since we're essentially still talking about join conditions.

Author: ok for me, adapted accordingly

If the [Logical Source]() of the [Triples Map]() that contains the [Referencing Object Map]() 
and the [Logical Source]() of the [Referencing Object Map]()'s [Parent Triples Map]() are not identical, 

> then the referencing object map must have at least one join condition.

Reviewer: ```suggestion
then the referencing object map MUST have at least one join condition.
```

> A [Logical Source]() is considered as identical to another [Logical Source]() 

Reviewer: ```suggestion
A [Logical Source]() is considered identical to another [Logical Source]() 
```

> when the set of objects at the end of the property paths starting with `rml:source` and starting with `rml:iterator` are identical. 

Reviewer: Why property paths? 

```suggestion
when
* the value of the source property (`rml:source`) of both logical sources is equal, and
* the value of the reference formulation property (`rml:referenceFormulation`) of both logical sources is equal, and
* the value of the iterator property (`rml:iterator`) of both logical sources is equal, or both logical sources do not specify the iterator property.
```

Author: @pmaria I couldn't think of another way to specify that also the descriptions of nested sources should be equals (so nested source descriptions can have different identifiers, but still have the same values for all nested properties (e.g. <LS1> and <LS2> are equal in the example, even is they have different identifiers.)

Reviewer 2: as discussed today:

```
when
* the value of the reference formulation property (`rml:referenceFormulation`) of both logical sources is equal,
* the value of the iterator property (`rml:iterator`) of both logical sources is equal, or both logical sources do not specify the iterator property, and
* the sub RDF graphs of the source property (`rml:source`) of both logical sources that only contain RML actionable properties of the source access descriptions are isomorph.
```

this last point means that each source access description needs to list its actionable properties

```
<LS1>
    a rml:LogicalSource; 
    rml:source <S1>;
    rml:referenceFormulation rml:JSONPath;
    rml:iterator "$.jsonpath.expression".
<S1> 
    a rml:Source, void:Dataset;
    void:dataDump <file:///data/dump.nt>. 
    
<LS2>
    a rml:LogicalSource; 
    rml:source <S2>;
    rml:referenceFormulation rml:JSONPath;
    rml:iterator "$.jsonpath.expression".
<S2> 
    a rml:Source, void:Dataset;
    void:dataDump <file:///data/dump.nt>.   
   
<LS3>
    a rml:LogicalSource; 
    rml:source <S1>;
    rml:referenceFormulation rml:JSONPath;
    rml:iterator "$.jsonpath.expression2".   
```

```
<LS1>
    a rml:LogicalSource; 
    rml:source [ a rml:Source, a csvw:Table
        csvw:url "/absolute/path/to/data.csv";
    ];
    rml:referenceFormulation rml:CSV.
    
<LS2>
    a rml:LogicalSource; 
    rml:source <S2>;
    rml:referenceFormulation rml:CSV.
    
<S2> 
   a rml:Source, a csvw:Table
   csvw:url "/absolute/path/to/data.csv".
  
<LS3>
    a rml:LogicalSource; 
    rml:source [ a rml:Source, a csvw:Table
        csvw:url "/relative/path/to/data.csv";
    ].
    rml:referenceFormulation rml:CSV.  
```

If the [Referencing Object Map]() has no join condition 

> (which is only allowed when the [Logical Source]() of the [Triples Map]() that contains the [Referencing Object Map]()

Reviewer: I would omit this sentence. It adds nothing to the earlier statement that unequal logical sources requires a referencing object map to have at least one join condition

and the [Logical Source]() of the [Referencing Object Map]()'s [Parent Triples Map]() are identical), a natural join is executed. 

> In reality this means that the [Logical Source]() is used in its original form when generating the related RDF triples.   

Reviewer: I would not introduce natural join as a term. It suffices to say that the child logical source is used.

> If the [Referencing Object Map]() has one or more join conditions, an inner join is executed. 

Reviewer: I would prefer not to introduce the term inner join if we can avoid it. In essence the join is explained in the next couple of sentences.

Author: ok for me, adapted accordingly

> The related RDF triples are generated using the [=n-ary Cartesian product=] 

Reviewer: If we can we should refer to a general description of generating triples. so that we don't have to repeat here that we use the n-ary cartesion product.

Author: @pmaria Can you please make a suggestion?

of the logical iteration of the [Logical Source]() of the [Triples Map]() that contains the [Referencing Object Map]()
and the logical iteration of the [Logical Source]() of the [Referencing Object Map]()'s [Parent Triples Map](), and
retaining only the combination of those logical iterations for which the values of the [Child Map]() and [Parent Map]() of each join condition are identical.

**NOTE**
If the [Referencing Object Map]() has no join condition and the [Logical Source]() of the [Triples Map]() that contains the [Referencing Object Map]()

> and the [Logical Source]() of the [Referencing Object Map]()'s [Parent Triples Map]() are not identical, the mapping engine MUST report an error.

Reviewer 3: Doesn't this break R2RML support? AFAIK you can join without a condition which results in joining everything from LS1 with everything of LS2 (Cartesian product)

Reviewer: No, R2RML states:

> If the [child query](https://www.w3.org/TR/r2rml/#dfn-child-query) and [parent query](https://www.w3.org/TR/r2rml/#dfn-parent-query) of a [referencing object map](https://www.w3.org/TR/r2rml/#dfn-referencing-object-map) are not identical, then the referencing object map must have at least one [join condition](https://www.w3.org/TR/r2rml/#dfn-join-condition).

Reviewer 3: AFAIK most engines allow this, thus violating the spec?

Reviewer: Yes, I would say that is a violation of the spec. We could of course argue about the usefulness of such behavior.

Reviewer 3: Hmmm this is of course not enforced in the shapes, it is kinda hard to do that I think.

Author: @pmaria @DylanVanAssche 
I am totally fine with the other solution: no join condition means cartesian product. I just want that a decision is taken on this matter, that decision is documented in the spec and that we all know that is not in line with the R2RML spec (what is the consequence?). 
I also believe that allowing 'no join condition' on any join is easier to define and implement. 
If we decide to move away from the R2RML spec anyway, I even wonder why we need the exception for 'same logical source'.  It would be much clearer if no join condition means cartesian product for any join. 
Since no such decisions were taken until now, I tried to write a PR in line with R2RML spec.

Reviewer 2: I would vote for the cartesian product: given that most engines implement it as such, it feels like it's the more intuitive interpretation of 'no join condition', and I'm all for increasing intuitivity! :). And we can see it as an extension of R2RML: cartesian allows you to do "more" than when you throw an error

Reviewer 4: FYI. R2RML --> if the queries of two different triples maps use different attributes for the generation of subject maps, then you must have a join condition (in other words, you must do a theta-join). When no join conditions are provided, then the rows of the child queries are used to populate both child and parent subject maps. In that sense, the no-join-conditions case _simulates_ a natural join. 

> If the referencing object map has no join condition:
> SELECT * FROM ({child-query}) AS tmp

This is sufficient, and the quote Pano mentioned is confusing. Testing the equivalence of two queries is "ignored" by the community. It was even the subject of a thread a while ago. Are `SELECT * FROM foo X` and `(SELECT * FROM foo Y)` and `(SELECT * FROM (SELECT * FROM foo))` identical or not?

> **NOTE**
> If the [Referencing Object Map]() has no join condition and the [Logical Source]() of the [Triples Map]() that contains the [Referencing Object Map]()
> and the [Logical Source]() of the [Referencing Object Map]()'s [Parent Triples Map]() are not identical, the mapping engine MUST report an error.

Reviewer: ```suggestion
<aside class="note">
If the [Referencing Object Map]() has no join condition and the [Logical Source]() of the [Triples Map]() that contains the [Referencing Object Map]()
and the [Logical Source]() of the [Referencing Object Map]()'s [Parent Triples Map]() are not identical, the mapping engine MUST report an error.
</aside>
```
